src/test/regress/expected/subselect.out

```diff
@@ -749,6 +749,36 @@ where s.i < 10 and (select val.x) < 110;
   10
 (17 rows)
 
+-- another variant of that (bug #16213)
+explain (verbose, costs off)
+select * from
+(values
+  (3 not in (select * from (values (1), (2)) ss1)),
+  (false)
+) ss;
+                  QUERY PLAN                  
+----------------------------------------------
+ Values Scan on "*VALUES*"
+   Output: "*VALUES*".column1
+   SubPlan 1
+     ->  Materialize
+           Output: "*VALUES*_1".column1
+           ->  Values Scan on "*VALUES*_1"
+                 Output: "*VALUES*_1".column1
+ Optimizer: Postgres query optimizer
```

Comment:
`Materialize` node here looks odd. This confirms the plan with costs:
```sql
explain (verbose)
select * from
(values
  (3 not in (select * from (values (1), (2)) ss1)),
  (false)
) ss;
                                 QUERY PLAN                                  
-----------------------------------------------------------------------------
 Values Scan on "*VALUES*"  (cost=0.00..0.03 rows=1 width=1)
   Output: "*VALUES*".column1
   SubPlan 1
     ->  Materialize  (cost=0.00..0.04 rows=1 width=4)
           Output: "*VALUES*_1".column1
           ->  Values Scan on "*VALUES*_1"  (cost=0.00..0.03 rows=1 width=4)
                 Output: "*VALUES*_1".column1
 Optimizer: Postgres query optimizer
```
`Materialize` node has more final cost than a whole plan that's absurd, as though it doesn't participate in execution pipeline.
@leskin-in if it's no bother could you check in planner logic the motivation to use `Materialize` here and such costs allocation?

Reply:
I am not sure of a precise reason for this, but I suppose this is normal for GPDB. See [`materialize_subplan()` method definition](https://github.com/arenadata/gpdb/blob/d28b705758eb743b5bcdf8aa2e4c658df33b6533/src/backend/cdb/cdbllize.c#L331-L335) - as far as I understand, it creates the `Materialize` node mentioned.

I have obtained this from a stacktrace, whose partial listing follows:
```
#0  cost_material (path=path@entry=0x7ffeca546c00, root=root@entry=0x7f545e4d3c68, input_startup_cost=0, input_total_cost=0.025000000000000001, tuples=2, width=4) at costsize.c:1640
#1  0x000055ec58938a1f in materialize_finished_plan (root=0x7f545e4d3c68, subplan=0x7f545e4e8a40) at createplan.c:5873
#2  0x000055ec58c612a9 in materialize_subplan (subplan=0x7f545e4e8a40, root=<optimized out>) at cdbllize.c:757
#3  ParallelizeSubplan (context=0x7ffeca547060, context=0x7ffeca547060, spExpr=0x7f545e4e7ef0) at cdbllize.c:757
#4  prescan_walker (node=0x7f545e4e7ef0, context=0x7ffeca547060) at cdbllize.c:855
...
```

Reply:
Yes, you are right: `Materialize` node is injected after execution of `materialize_subplan()` function. This is done inside `ParallelizeSubplan()` routine that tries our `ValuesScan` subplan to make distributed. The motivation is described in the commit fb7247b93b:
```
    Consider below query:
    
            INSERT INTO foo SELECT f1+10, f2, f3+99 FROM foo
              RETURNING *, f1+112 IN (SELECT q1 FROM int8_tbl) AS subplan;
    
    We used to generate a plan like this:
    
            Explicit Gather Motion 3:1  (slice2; segments: 3)
              ->  Insert
                    ->  Seq Scan on foo
                    SubPlan 1  (slice2; segments: 3)
                      ->  Gather Motion 3:1  (slice1; segments: 1)
                            ->  Seq Scan on int8_tbl
    
    A gather motion is used for the subplan, which is wrong and will cause a
    runtime error.
    
    A correct plan is like below:
    
            Explicit Gather Motion 3:1  (slice2; segments: 3)
              ->  Insert
                    ->  Seq Scan on foo
                    SubPlan 1  (slice2; segments: 3)
                      ->  Materialize
                            ->  Broadcast Motion 3:3  (slice1; segments: 3)
                                  ->  Seq Scan on int8_tbl
```
That is, `Materialize` node as upper one for a subplan is intended to aggregate results from any sub-motions. This confirms the [comment](https://github.com/arenadata/gpdb/blob/d28b705758eb743b5bcdf8aa2e4c658df33b6533/src/backend/cdb/cdbllize.c#L746-L758) above `materialize_subplan()` call.
But our subplan is not distributed one and cost addition after `Materialize` node injecting to a whole plan doesn't occur. I think these questions might be raised in a new ticket.